fsw/c-blinky/main.c

```diff
@@ -0,0 +1,238 @@
+    (void*) &_stack_end,
+    Reset_Handler,
+    Default_Handler,
+    Default_Handler,
+    Default_Handler,
+    Default_Handler,
+    Default_Handler,
+    0,
+    0,
+    0,
+    0,
+    Default_Handler,
+    Default_Handler,
+    0,
+    Default_Handler,
+    SysTick_Handler,
+};
+
+static inline void rtt_init(void) {
+    _SEGGER_RTT.aUp[0] = (RTT_BUFFER_UP) {
+        .sName        = "Terminal",
+        .pBuffer      = _acUpBuffer,
+        .SizeOfBuffer = sizeof(_acUpBuffer),
+    };
+}
+
+static inline void rtt_write(const char* s) {
+    if (!s) return;
+    RTT_BUFFER_UP* pRing = &_SEGGER_RTT.aUp[0];
+    while (*s) {
+        pRing->pBuffer[pRing->WrOff] = *s++;
+        pRing->WrOff                 = (pRing->WrOff + 1) % pRing->SizeOfBuffer;
+    }
+}
+
+static void system_init(void) {
+    PWR_CR3 &= ~PWR_CR3_BYPASS;
+    PWR_CR3 &= ~PWR_CR3_SDEN;
+    PWR_CR3 |= PWR_CR3_LDOEN;
+    PWR_D3CR |= PWR_D3CR_VOS_SCALE1;
+
+    U32 vos_timeout = 1000000;
+    while (!(PWR_CSR1 & PWR_CSR1_ACTVOSRDY) && vos_timeout--);
+
+    FLASH_ACR = FLASH_ACR_LATENCY_4WS | FLASH_ACR_PRFTEN | FLASH_ACR_ICEN | FLASH_ACR_DCEN;
+}
+
+static void clock_config(void) {
+    RCC_CR |= RCC_CR_HSEBYP | RCC_CR_HSEON;
+    U32 hse_timeout = 1000000;
+    while (!(RCC_CR & RCC_CR_HSERDY) && hse_timeout--);
+
+    RCC_D1CFGR = RCC_D1CFGR_D1CPRE_DIV1 | RCC_D1CFGR_HPRE_DIV2;
+    RCC_D2CFGR = RCC_D2CFGR_D2PPRE1_DIV2 | RCC_D2CFGR_D2PPRE2_DIV2;
+    RCC_D3CFGR = RCC_D3CFGR_D3PPRE_DIV2;
+
+    RCC_PLL1CFGR = RCC_PLL1CFGR_PLL1SRC_HSE | RCC_PLL1CFGR_PLL1PEN | RCC_PLL1CFGR_PLL1QEN |
+                   ((PLL1_M - 1) << 4);
+
+    RCC_PLL1DIVR = ((PLL1_P - 1) << 9) | ((PLL1_Q - 1) << 16) | ((PLL1_N - 1) << 0);
+
+    RCC_CR |= RCC_CR_PLL1ON;
+    U32 pll_timeout = 1000000;
+    while (!(RCC_CR & RCC_CR_PLL1RDY) && pll_timeout--);
+
+    RCC_CFGR           = (RCC_CFGR & ~0x7) | RCC_CFGR_SW_PLL1;
+    U32 switch_timeout = 1000000;
+    while ((RCC_CFGR & 0x38) != RCC_CFGR_SWS_PLL1 && switch_timeout--);
+}
+
+static void gpio_config(void) {
+    RCC_AHB4ENR |= RCC_AHB4ENR_GPIOEEN;
+    // Configure PE5 (red LED) as output
+    GPIOE_MODER &= ~(3 << (RED_LED_PIN * 2));
+    GPIOE_MODER |= (1 << (RED_LED_PIN * 2));
+    GPIOE_OTYPER &= ~(1 << RED_LED_PIN);
+    GPIOE_OSPEEDR &= ~(3 << (RED_LED_PIN * 2));
+}
+
+static void systick_init(void) {
+    SYSTICK_RVR = 400000 - 1;
+    SYSTICK_CVR = 0;
+    SYSTICK_CSR = SYSTICK_CSR_CLKSOURCE | SYSTICK_CSR_TICKINT | SYSTICK_CSR_ENABLE;
+}
+
+void SysTick_Handler(void) {
+    systick_ms++;
+}
+
+static void delay_ms(U32 ms) {
+    U32 start = systick_ms;
+    while ((systick_ms - start) < ms);
+}
+
+void Reset_Handler(void) {
+    U32* src  = &_sidata;
+    U32* dest = &_sdata;
+
+    while (dest < &_edata) *dest++ = *src++;
+    dest = &_sbss;
+    while (dest < &_ebss) *dest++ = 0;
+
+    main();
+    while (1);
+}
+
+void Default_Handler(void) {
+    while (1);
```

Comment:
Do you actually want to loop forever as the default behavior for all interrupts? I think generally you would want to trigger a breakpoint so the debugger could figure out what happened? Or clear the interrupt register and drop back to the previous state